packages/@tailwindcss-upgrade/src/template/codemods/legacy-classes.ts

```diff
@@ -0,0 +1,110 @@
+import { __unstable__loadDesignSystem } from '@tailwindcss/node'
+import path from 'node:path'
+import url from 'node:url'
+import type { Config } from 'tailwindcss'
+import type { DesignSystem } from '../../../../tailwindcss/src/design-system'
+import { printCandidate } from '../candidates'
+
+const __filename = url.fileURLToPath(import.meta.url)
+const __dirname = path.dirname(__filename)
+
+const LEGACY_CLASS_MAP = {
+  shadow: 'shadow-sm',
+  'shadow-sm': 'shadow-xs',
+  'shadow-xs': 'shadow-2xs',
+
+  'inset-shadow': 'inset-shadow-sm',
+  'inset-shadow-sm': 'inset-shadow-xs',
+  'inset-shadow-xs': 'inset-shadow-2xs',
+
+  'drop-shadow': 'drop-shadow-sm',
+  'drop-shadow-sm': 'drop-shadow-xs',
+
+  rounded: 'rounded-sm',
+  'rounded-sm': 'rounded-xs',
+
+  blur: 'blur-sm',
+  'blur-sm': 'blur-xs',
+}
+
+const THEME_KEYS = {
+  shadow: '--shadow',
+  'shadow-sm': '--shadow-sm',
+  'shadow-xs': '--shadow-xs',
+  'shadow-2xs': '--shadow-2xs',
+
+  'drop-shadow': '--drop-shadow',
+  'drop-shadow-sm': '--drop-shadow-sm',
+  'drop-shadow-xs': '--drop-shadow-xs',
+
+  rounded: '--radius',
+  'rounded-sm': '--radius-sm',
+  'rounded-xs': '--radius-xs',
+
+  blur: '--blur',
+  'blur-sm': '--blur-sm',
+  'blur-xs': '--blur-xs',
+}
+
+const SEEDED = new WeakSet<DesignSystem>()
+
+export async function legacyClasses(
+  designSystem: DesignSystem,
+  _userConfig: Config,
+  rawCandidate: string,
+): Promise<string> {
+  // Prepare design system with the unknown legacy classes
+  if (!SEEDED.has(designSystem)) {
+    for (let old in LEGACY_CLASS_MAP) {
+      designSystem.utilities.static(old, () => [])
+    }
+    SEEDED.add(designSystem)
+  }
+
+  let defaultDesignSystem = await __unstable__loadDesignSystem('@import "tailwindcss";', {
+    base: __dirname,
+  })
```

Comment:
Should move this outside of the function scope, it's doing this for every candidate otherwise